bin/run_docker.py

```diff
@@ -267,8 +268,55 @@ def mount_volumes() -> dict:
     return volumes
 
 
+def monitor_container(
+    container: docker.models.containers.Container,
+    timeout: Union[int, float],
+    poll_interval: Union[int, float],
+    elapsed_time: int = 0,
+):
+    """
+    Recursively monitor the status of a Docker container until it finishes,
+    or the timeout is reached, at which point the container is killed.
+
+    Arguments:
+        container: The Docker container subject to monitor.
+        timeout: Maximum duration to monitor the container (in minutes) before forced shutdown.
+        poll_interval: Time to wait between status checks (in minutes).
+        elapsed_time: The elapsed time already spent monitoring (in minutes).
+    """
+
+    # Check if the elapsed time has reached or exceeded the timeout
+    if elapsed_time >= timeout:
```

Comment:
Does this always give each submission exactly the same maximum amount of time to execute?

Reply:
Maximum time would probably end up being `container_timeout` + `poll_interval` - assume a situation where polling executes a second before timeout. The container would get an extra 10 minutes, but would eventually time out.

Reply:
Yes, at most the amount of time the container can run will be up to `timeout` + `poll_interval` - 1 (unit: minutes)

----
e.g.
* A docker container runs for 14 minutes 59 seconds
* the timeout limit is 11 mins
* the poll interval is 5 mins

At min=0
* container status = `"running"`
* `elapsed_time` = 0, AKA under `timeout`, so it keeps running
* `elapsed_time` becomes `elapsed_time` + 5 min
* the monitor sleeps for 5 mins

At min=5
* container status = `"running"`
* `elapsed_time` = 5, AKA under `timeout`, so it keeps running
* `elapsed_time` becomes `elapsed_time` + 5 min
* the monitor sleeps for 5 mins

At min=10
* container status = `"running"`
* `elapsed_time` = 10, AKA under `timeout`, so it keeps running
* `elapsed_time` becomes `elapsed_time` + 5 min
* the monitor sleeps for 5 mins

At min=15
* container status = `"exited"`, the function returns an empty string
----

So a docker submission that extends beyond the allotted time limit can definitely sneak through without a timeout error, which is not ideal, but based on the scenarios I've explored, all cases beyond `timeout` + `poll_interval` - 1 will fail. Enforcing the timeout while also accommodating for `timeout - poll_interval < docker run time < timeout` edge cases is something I can explore in a future ticket.

Reply:
I think the more immediate (and simple) way is to do something like a min function where the `time.sleep` will sleep either the `poll_interval` or the remaining time to reach `timeout` if `poll_interval` + elapsed time is > `timeout`

Reply:
I was actually thinking of removing `poll_interval` as a user-fed parameter and calculating it programmatically based on the `timeout` that's fed. I'd like to give this some more thought before pushing anything into `main` related to this. The default `timeout` and `poll_interval` work for PEGS since the `poll_interval` is much smaller than the `timeout`.

Reply:
Thanks @BryanFauble - I originally thought to remove the `poll_interval` as an argument and compute it behind the scenes to avoid things like overlap in the final poll interval, but I think it's best to leave it exposed as a parameter in case we need to modify it for troubleshooting or the like. I agree your suggestion is the most straight-forward way to tackle this and I've implemented it in https://github.com/Sage-Bionetworks-Workflows/nf-synapse-challenge/pull/26/commits/d040b746237a3ac2bc83fa503232567c50fee263.

Here are some test results from running the script locally:

test 1) `poll_interval` = 3 mins (longer than the `timeout`)
```
(synapseclient) (synapseclient) ip-10-0-0-173:nf-synapse-challenge jmedina$ python bin/run_docker.py 9744669 2 3 50 

UPGRADE AVAILABLE

A more recent version of the Synapse Client (4.2.0) is available. Your version (4.1.1) can be upgraded by typing:
    pip install --upgrade synapseclient

Python Synapse Client version 4.2.0 release notes

https://python-docs.synapse.org/news/

Running container... docker.synapse.org/syn51317219/runs_1d5min@sha256:1136a32f2bfbeb38be8667413cbde052cc88f063e512a519d66051592854a399
Container still running. Checking again in 2.0 minutes.
Time spent monitoring container (minutes): 2.0
Container run will shut down after 2.0 minute(s).
Original log message exceeds 50 Kb. Truncating...
Truncated log file created: 9744669_docker.log
File renamed successfully.
Original file:  /Users/jmedina/Documents/forks/nf-synapse-challenge/output/predictions.csv
Renamed file:  /Users/jmedina/Documents/forks/nf-synapse-challenge/output/9744669_predictions.csv
```

test 2) `poll_interval` = 0.5 mins
```
(synapseclient) (synapseclient) ip-10-0-0-173:nf-synapse-challenge jmedina$ python bin/run_docker.py 9744669 2 0.5 50 

UPGRADE AVAILABLE

A more recent version of the Synapse Client (4.2.0) is available. Your version (4.1.1) can be upgraded by typing:
    pip install --upgrade synapseclient

Python Synapse Client version 4.2.0 release notes

https://python-docs.synapse.org/news/

Running container... docker.synapse.org/syn51317219/runs_1d5min@sha256:1136a32f2bfbeb38be8667413cbde052cc88f063e512a519d66051592854a399
Container still running. Checking again in 0.5 minutes.
Container still running. Checking again in 0.5 minutes.
Container still running. Checking again in 0.5 minutes.
Time spent monitoring container (minutes): 1.5
Container run will shut down after 2.0 minute(s).
Container still running. Checking again in 0.5 minutes.
Time spent monitoring container (minutes): 2.0
Container run will shut down after 2.0 minute(s).
Original log message exceeds 50 Kb. Truncating...
Truncated log file created: 9744669_docker.log
File renamed successfully.
Original file:  /Users/jmedina/Documents/forks/nf-synapse-challenge/output/predictions.csv
Renamed file:  /Users/jmedina/Documents/forks/nf-synapse-challenge/output/9744669_predictions.csv
```

test 3) `poll_interval` = 1.5 mins, `timeout` = 2 mins, which means the last poll would exceed `timeout`
```
(synapseclient) (synapseclient) ip-10-0-0-173:nf-synapse-challenge jmedina$ python bin/run_docker.py 9744669 2 1.5 50 

UPGRADE AVAILABLE

A more recent version of the Synapse Client (4.2.0) is available. Your version (4.1.1) can be upgraded by typing:
    pip install --upgrade synapseclient

Python Synapse Client version 4.2.0 release notes

https://python-docs.synapse.org/news/

Running container... docker.synapse.org/syn51317219/runs_1d5min@sha256:1136a32f2bfbeb38be8667413cbde052cc88f063e512a519d66051592854a399
Container still running. Checking again in 1.5 minutes.
Time spent monitoring container (minutes): 1.5
Container run will shut down after 2.0 minute(s).
Container still running. Checking again in 0.5 minutes.
Time spent monitoring container (minutes): 2.0
Container run will shut down after 2.0 minute(s).
Original log message exceeds 50 Kb. Truncating...
Truncated log file created: 9744669_docker.log
File renamed successfully.
Original file:  /Users/jmedina/Documents/forks/nf-synapse-challenge/output/predictions.csv
Renamed file:  /Users/jmedina/Documents/forks/nf-synapse-challenge/output/9744669_predictions.csv
```